### DISCUSSION TITLE
Get link text by cursor position

### DISCUSSION THREAD

**Discussion by Dragonite24** [ueberdosis/tiptap#4716] [General Questions]
I want to do the following functionality:
There is text and a link in it, having focus on the text with the link (the text is not selected, just focus) when you click on some button, I want to be able (in a separate popup component) to change the current link **and text**.

In the pictures, I want to get out of this: 
![image](https://github.com/ueberdosis/tiptap/assets/60312801/81783fd2-fa06-4396-9335-ffa645261807)
_(...some edit in my popup)_ to this:
![image](https://github.com/ueberdosis/tiptap/assets/60312801/4f069976-1509-48e2-be03-cf3a915df5f2)

There are no problems with changing the link. **But how to get text of link by cursor position**? and is it possible?

**Comment by flowt-au** [comment]
Did you ever work this out? I need the same thing.

**Comment by jrief** [comment]
I use the following snippet to set the link with some content:

```javascript
const selection = editor.view.state.selection;
editor.chain().focus()
  .extendMarkRange('link')
  .setLink({href: 'http://www.example.org'})
  .insertContentAt({from: selection.from, to: selection.to}, 'EditTextAndLink')
  .run();
```

**Reply by andersr** [reply]
@jrief  This is useful info but I think is doing the opposite of what the OP is asking for which is _getting_, not setting the text link.

I'm also trying to access the link text.  I've made numerous attempts at trying to get this but no luck so far.  Here are some things I've tried:

```javascript
const attr = editor.getAttributes("link");
```
This gets the attributes of the link itself, but unfortunately does not include the link text, which I guess is a child node(?)

```javascript
   editor .chain() .focus().extendMarkRange("link").command(({ tr }) => {
      // tried consoling out the tr to see if it included the text but no luck.
 }).run();
```

Have read through both the TipTap and ProseMirror docs but so far have not been able to find anything.  Any insights would be appreciated!

**Reply by nperez0111** [reply]
@andersr you were very close, `editor.chain().focus().extendMarkRange('link')` [sets the selection to be as wide as the mark is applied](https://github.com/ueberdosis/tiptap/blob/b941eea6daba09d48a5d18ccc1b9a1d84b2249dd/packages/core/src/commands/extendMarkRange.ts#L42-L44)

So if you console log `tr.selection.content()` that is the slice of the content that this range applies to.

Hope this is helpful!

**Reply by andersr** [reply]
Thanks @nperez0111. However, that did not work for me. 

This is my code:
```javascript
   editor.chain().focus().extendMarkRange("link").command(({ tr }) => {
    console.log("tr: ", tr.selection.content());
    return true;
 }) .run();
```

This is what I see in the console (maybe the link text is in there somewhere?):
![Screenshot 2024-08-09 at 11 14 55 AM](https://github.com/user-attachments/assets/15e21324-5e76-46ac-a492-51a6272389a8)

**Reply by nperez0111** [reply]
The text should be in that sub `content` somewhere, though it looks like it copied the whole bulletList which is unexpected

**Comment by andersr** [comment]
This worked for me:

```javascript
 const { from } = editor.view.state.selection;
 const linkNode = editor.state.doc.nodeAt(from);
 console.log("linkNode: ", linkNode?.textContent);
```
(I got that based on a partially correct answer from ChatGPT that I was able to make work.)

**Reply by nperez0111** [reply]
Yea that can work too. Just not as generalizable to other mark types since they may not only contain text. But if it works, it works
